node/impl/eth/tipsetresolver.go

```diff
@@ -21,117 +21,153 @@ import (
 	head := tsr.cs.GetHeaviestTipSet()
-	// for Ethereum APIs we focus on the message inclusion tipset so we have to -1 here, in contrast
-	// to the Filecoin native height selectors
-	height := head.Height() - 1 - policy.ChainFinality
-	if tag == "safe" {
+	height := head.Height()
+
+	// Default safe distance (used for v2 APIs, or v1 APIs when F3 is finalizing)
+	safeDistance := buildconstants.SafeHeightDistance
+
+	// For v1 backward compatibility: if not using F3 for finality AND F3 is not actually finalizing,
+	// use the older safe distance.
+	// ethtypes.SafeEpochDelay can be removed in the future when F3 is always finalizing.
+	if !tsr.useF3ForFinality {
+		f3Finalizing := false
 		if tsr.f3 != nil {
-			// for v2 APIs we'll be consistent with the non-Ethereum APIs
-			height = head.Height() - 1 - buildconstants.SafeHeightDistance
-		} else {
-			// for v1 backward compatibility we use the EC finality height
-			// NOTE: this may change in the future to be buildconstants.SafeHeightDistance
-			height = head.Height() - 1 - ethtypes.SafeEpochDelay
+			// Attempt to get manifest; ignore errors (treat as not finalizing if error occurs or f3 is nil)
+			manifest, err := tsr.f3.GetManifest(ctx)
+			if err == nil && manifest != nil {
+				f3Finalizing = manifest.EC.Finalize
+			}
+		}
+		if !f3Finalizing {
+			safeDistance = ethtypes.SafeEpochDelay
 		}
 	}
-	if height < 0 {
-		height = 0
-	}
-	return tsr.cs.GetTipsetByHeight(ctx, height, head, true)
+
+	safeHeight := max(0, height-safeDistance)
+	return tsr.cs.GetTipsetByHeight(ctx, safeHeight, head, true)
 }
 
-var handledLabels = []string{"earliest", "pending", "latest", "safe", "finalized"}
+func (tsr *tipSetResolver) getFinalizedECTipSet(ctx context.Context) (*types.TipSet, error) {
+	head := tsr.cs.GetHeaviestTipSet()
+	height := max(0, head.Height()-policy.ChainFinality)
+	return tsr.cs.GetTipsetByHeight(ctx, height, head, true)
+}
 
-func (tsr *tipSetResolver) getTipSetByLabel(ctx context.Context, label string) (*types.TipSet, error) {
+func (tsr *tipSetResolver) getTipSetByTag(ctx context.Context, tag string) (*types.TipSet, error) {
 	head := tsr.cs.GetHeaviestTipSet()
-	switch label {
-	case "earliest":
-		return nil, xerrors.New("block param \"earliest\" is not supported")
-	case "pending":
+	switch tag {
+	case ethtypes.BlockTagEarliest:
+		return nil, xerrors.Errorf(`block param "%s" is not supported`, ethtypes.BlockTagEarliest)
+	case ethtypes.BlockTagPending:
 		return head, nil
-	case "latest":
+	case ethtypes.BlockTagLatest:
 		parent, err := tsr.cs.GetTipSetFromKey(ctx, head.Parents())
 		if err != nil {
 			return nil, xerrors.New("failed to get parent tipset")
 		}
 		return parent, nil
-	case "safe":
+	case ethtypes.BlockTagSafe:
 		if tsr.f3 != nil {
-			return tsr.getTipSetByTag(ctx, "safe")
+			ts, err := tsr.getSafeF3TipSet(ctx)
+			if err != nil && tsr.useF3ForFinality {
+				return nil, err
```

Comment:
IMO, "safe" should always fallback on EC even if we're using F3 for finality, but I don't have strong opinions on this.

Reply:
Deferring further discussion to https://github.com/filecoin-project/lotus/issues/13060